File: snapshot_array.py
class SnapshotArray:

    def __init__(self, length: int):
        self.snapshot = {i: {0: 0} for i in range(length)}
        self.snap_id = 0

    def set(self, index: int, val: int) -> None:
        self.snapshot[index][self.snap_id] = val

    def snap(self) -> int:
        self.snap_id += 1
        return self.snap_id - 1

    def get(self, index: int, snap_id: int) -> int:
        while snap_id > 0 and snap_id not in self.snapshot[index]:
            snap_id -= 1
        return self.snapshot[index][snap_id]

File: snapshot_array.cpp
class SnapshotArray {
public:
    unordered_map<int, unordered_map<int, int>> database;
    unordered_map<int, int> cur_data;
    int snap_id;
    SnapshotArray(int length) {
      snap_id = -1;
    }
    
    void set(int index, int val) {
      cur_data[index] = val;
    }
    
    int snap() {
      snap_id++;
      database[snap_id] = cur_data;
      return snap_id;
    }
    
    int get(int index, int snap_id) {
      return database[snap_id][index];
    }
};
/**
 * Your SnapshotArray object will be instantiated and called as such:
 * SnapshotArray* obj = new SnapshotArray(length);
 * obj->set(index,val);
 * int param_2 = obj->snap();
 * int param_3 = obj->get(index,snap_id);
 */
